File: lib/js/rocketio.js
var RocketIO = function(opts){
  new EventEmitter().apply(this);
  if(typeof opts === "undefined" || opts === null) opts = {};
  this.type = opts.type || null; // "comet" or "websocket"
  this.session = opts.session || null;
  this.channel = null;
  if(typeof opts.channel !== "undefined" && opts.channel !== null){
    this.channel = ""+opts.channel;
  }
  var setting = {};
  this.io = null;
  var self = this;
  var ws_close_timer = null;
  self.on("__connect", function(session_id){
    self.session = session_id;
    self.io.push("__channel_id", self.channel);
    self.emit("connect");
  });

  this.connect = function(url){
    if(typeof url === "string"){
      $.getJSON(url+"/rocketio/settings", function(res){
        setting = res;
        connect_io();
      });
      return self;
    }
    else{
      return connect_io();
    }
  };

  var connect_io = function(){
    self.io = function(){
      if(self.type === "comet") return;
      if(typeof WebSocketIO !== "function") return;
      var io = new WebSocketIO();
      if(typeof setting.websocket === "string") io.url = setting.websocket;
      io.session = self.session;
      return io.connect();
    }() || function(){
      if(typeof CometIO !== "function") return;
      var io = new CometIO();
      if(typeof setting.comet === "string") io.url = setting.comet;
      io.session = self.session;
      return io.connect();
    }();
    if(typeof self.io === "undefined"){
      setTimeout(function(){
        self.emit("error", "WebSocketIO and CometIO are not available");
      }, 100);
      return self;
    };
    if(self.io.url.match(/^ws:\/\/.+/)) self.type = "websocket";
    else if(self.io.url.match(/cometio/)) self.type = "comet";
    else self.type = "unknown";
    self.io.on("*", function(event_name, args){
      if(event_name === "connect") event_name = "__connect";
      self.emit(event_name, args);
    });
    ws_close_timer = setTimeout(function(){
      self.close();
      self.type = "comet";
      connect_io();
    }, 3000);
    self.once("connect", function(){
      if(ws_close_timer) clearTimeout(ws_close_timer);
      ws_close_timer = null;
    });
    return self;
  };

  this.close = function(){
    self.io.close();
  };

  this.push = function(type, data){
    self.io.push(type, data);
  };
};

File: lib/sinatra-rocketio/rocketio.rb
EventEmitter.apply Sinatra::RocketIO

module Sinatra
  module RocketIO

    def self.push(type, data, opt={})
      if opt.include? :to and opt[:to].kind_of? Array
        opt[:to].each do |to|
          push type, data, :to => to
        end
      elsif opt.include? :channel
        channels.select{|session, channel|
          channel == opt[:channel].to_s
        }.each{|session, channel|
          push type, data, :to => session
        }
      else
        if options[:websocket]
          Sinatra::WebSocketIO.push type, data, opt
        end
        if options[:comet]
          Sinatra::CometIO.push type, data, opt
        end
      end
    end

    def self.sessions
      {
        :websocket => Sinatra::WebSocketIO.sessions.keys,
        :comet => Sinatra::CometIO.sessions.keys
      }
    end

  end
end

Sinatra::RocketIO.once :start do
  if options[:comet]
    Sinatra::CometIO.on :* do |event_name, *args|
      event_name = :__connect if event_name == :connect
      session_id = args.size > 1 ? args[1] : args[0]
      address = sessions.include?(session_id) ? sessions[session_id][:remote_addr] : nil
      info = Sinatra::RocketIO::ClientInfo.new(:session => session_id, :address => address, :channel => Sinatra::RocketIO.channels[session_id], :type => :comet)
      if args.size > 1
        Sinatra::RocketIO.emit event_name, args[0], info
      else
        Sinatra::RocketIO.emit event_name, info
      end
    end
  end
  if options[:websocket]
    Sinatra::WebSocketIO.on :* do |event_name, *args|
      event_name = :__connect if event_name == :connect
      session_id = args.size > 1 ? args[1] : args[0]
      address = sessions.include?(session_id) ? sessions[session_id][:remote_addr] : nil
      info = Sinatra::RocketIO::ClientInfo.new(:session => session_id, :address => address, :channel => Sinatra::RocketIO.channels[session_id], :type => :websocket)
      if args.size > 1
        Sinatra::RocketIO.emit event_name, args[0], info
      else
        Sinatra::RocketIO.emit event_name, info
      end
    end
  end
end
